### ISSUE TITLE
In markdown links, automatically convert ".md" into ".html"?

### ISSUE THREAD

**Issue by calen-legaspi** [concordion/concordion#319]
In markdown, I'd like the raw files be able to link to each other like this:

`[link to other spec](OtherSpec.md)` 

However, when I do that, the generated HTML links are broken because they end up like this:

`<a href="OtherSpec.md">link to other spec</a>`

So what I end up doing is writing the markdown links like this:

`[link to other spec](OtherSpec.html)`

but now the links in the raw markdowns are broken.

I think it would be convenient if we could write proper markdown links, and then the spec generator would simply convert all the ".md" into ".html"...

`<a href="OtherSpec.html">link to other spec</a>`

Would it be alright to see this in a future build? Maybe there would be a scan first if the md files actually exist before doing the conversion, just in case the author really wants to link to an external URL with a ".md" extension.

**Comment by nigelcharman** [comment]
It might be best if this was an explicit choice rather than implicit. Would it work for you if we added a new command for this?

For example:

`[link to other spec](OtherSpec.md "c:relink")`

This would be an easy option to implement.

**Comment by calen-legaspi** [comment]
I suppose that works. But would there be a case where a person would link to a .md file and not mean to link to another page in the spec? I'm guessing it's highly unlikely, which is why I'm of the opinion the default behaviour should be converting "X.md" to "X.html" if X.md exists. 

By the way, I've got a little free time. I could write a patch that implements this. Just let me know what packages and classes you think I should be editing so I don't clobber the wrong parts?

**Comment by nigelcharman** [comment]
Just checking you know that the Run Command has this behaviour? The Run Command does cache called specs so will only execute them once.

Assuming you still want to implement this feature rather than using the Run Command:

I guess it's unlikely that a **relative** URL link would be to a `.md` file that's not part of the spec. That would need the `.md` file to be copied as a resource, or for the spec to be copied somewhere where the `.md` file already exists. 

However, if the link has an **absolute** URL it's unlikely to be part of the spec, so shouldn't be converted.

Would you build a specification for this new behaviour and then post a link on our Google Group to see if anybody objects? I'm happy to review the spec before you post it.

That would be great if you have time to implement it too. Here's some notes on how it might work:

- To support a range of different input formats (such as Markdown or Excel), Concordion has a `SpecificationConverter` interface that converts from the input format to HTML. You can also mix input formats, so you could have an HTML spec with a link to a `.md` spec that needs updating to `.html`.
- I suggest we implement it in the HTML Parser, rather than the MarkdownParser. This will then apply the same behaviour to any input format.
- It might need to go in the `DocumentParser::parse(Document document, Resource resource)` method. Once we get past this method, we're only working on Commands. If so, you'd need to work through the whole Document looking for `a` elements and updating any applicable `href`s.
- This method currently renames links in `RunCommand.java`:
```
    /* For non-html format specs, update the href to reference the html output */
    private void updateHrefSuffix(Element element, String href) {
        if (!(href.endsWith(".html"))) {
            int dot = href.lastIndexOf(".");
            if (dot > 0) {
                String modifiedHref = href.substring(0, dot) + ".html";
                element.removeAttribute("href");
                element.addAttribute("href", modifiedHref);
            }
        }
    }
```
- Alternatively, you could see if you could implement it in the XOM `Builder` which is called from `XMLBuilder.parse(InputStream inputStream, String sourceDescription)`, possibly using an `XMLFilter`. This would be more performant, but I'm not sure if it's even possible. I haven't worked on this part of the codebase.

**Comment by calen-legaspi** [comment]
Sure, I'll draft a spec for your review. Shall I email the draft to your personal email address?

Thanks for the advice on how to implement the patch!

**Comment by calen-legaspi** [comment]
Hi, Nigel!

I just realized that Concordion won't generate HTML from a markdown file unless there was a corresponding fixture. I was thinking it would be nice to be able to create pages in markdown that weren't specs, just plain-old pages that would contain supplementary info. Does it make sense to change the behavior of Concordion so it converts markdown files to HTML even if they aren't specs?

**Comment by nigelcharman** [comment]
Hi, applying an empty fixture when one doesn't exist is a feature that's been often requested. See issues #5 and #24. It would be worth having a think about implementing this, with the benefit that the output would have the same styling as Concordion specs. Do you want to tackle this first? I'd suggest creating a new issue for this behaviour.

**Comment by calen-legaspi** [comment]
Sure, makes sense to tackle this first. Can I change the requirement a bit such that we can create markdown pages that aren't specs? Right now, if I were to create a page that wasn't a spec, I'd need to do it in HTML, since Concordion wouldn't convert a markdown page that wasn't a spec.

**Comment by nigelcharman** [comment]
Yes please. Ideally it would extend out to cover any other formats that have a SpecificationConverter, but let's see where we land with Markdown and HTML files to start with. 
Note the extensions we support by default is defined in the ConcordionBuilder class as:
```
        withSpecificationType("html", null);
        withSpecificationType("xhtml", xhtmlConverter);
        withSpecificationType("md", markdownConverter);
        withSpecificationType("markdown", markdownConverter);
```

**Comment by calen-legaspi** [comment]
Great! I'll create the issue.
